internal/transport/http2_client.go

```diff
@@ -1006,24 +1008,29 @@ func (t *http2Client) Close(err error) {
 		t.kpDormancyCond.Signal()
 	}
 	t.mu.Unlock()
-	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the
-	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY.
+	var st *status.Status
+	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the connection.
+	// See https://httpwg.org/specs/rfc7540.html#GOAWAY. It also waits for loopyWriter to
+	// be closed with a timer to avoid the indefinite blocking.
 	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte("client transport shutdown"), closeConn: err})
-	<-t.writerDone
+	timer := time.NewTimer(GoAwayLoopyWriterTimeout)
```

Comment:
I would use `time.After` for simplicity purposes. See: https://pkg.go.dev/time#After.

If you feel that efficiency (in terms of the timer object being garbage collected) is a concern, then you should arrange for the timer to be `Close`ed.